adsdata/metrics.py

```diff
@@ -261,261 +265,149 @@ def row_view_to_metrics(self, passed_row_view, nonbib_db_conn, row_view_schema='
             return False
 
-        if type(m1[fieldname]) != type(m2[fieldname]):
+        if type(v1) != type(v2):
+            metrics_logger.warn('{} field type mismatch: {} {}'.format(type(v1), type(v2)))
             return True
 
-        if m1[fieldname] == None and m2[fieldname] == None:
+        if v1 == None and v2 == None:
             return False
 
         if fieldname in ('downloads', 'reads'):
             # only last value may be different to account for slightly older test data
-            t1 = m1[fieldname][:-1]
-            t2 = m2[fieldname][:-1]
-            for v1,v2 in zip(t1, t2):
-                if v1 != v2:
-                    print fieldname, v1, v2
+            t1 = v1[:-1]
+            t2 = v2[:-1]
+            for x1,x2 in zip(t1, t2):
+                if x1 != x2:
+                    metrics_logger.warn('{} {} arrays differ on not last element: {} {}'.format(bibcode, fieldname, x1, x2))
                     return True;
-            t1 = m1[fieldname][-1]
-            t2 = m2[fieldname][-1]
+            t1 = v1[-1]
+            t2 = v2[-1]
             delta = abs(t1 - t2)
-            limit = max(5, v1 * .15)
+            limit = max(5, t1 * .15)
             if delta >  limit:
-                print fieldname, v1, v2
+                metrics_logger.warn('{} {} last element of arrays differ: {} {}, threshold: {}'.format(bibcode, fieldname, t1, t2, limit))
                 return True
             return False
 
         # if value is an array, compare individual elements
-        if isinstance(m1[fieldname], list):
-            delta = abs(len(m1[fieldname]) - len(m2[fieldname]))
+        if isinstance(v1, list):
+            limit = 3
+            delta = abs(len(v1) - len(v2))
             if delta > 0 and delta < 3:
-                print 'warning: list length', len(m1[fieldname]), len(m2[fieldname])
-            elif delta >= 3:
+                metrics_logger.info('{} {} array elements nearly match, lengh delta: {}'.format(bibcode, fieldname, delta))
+            elif delta >= limit:
+                metrics_logger.warn('{} {} array elements do not match, length delta {}, threshold: {}'.format(bibcode, fieldname, delta, limit))
                 return True
 
             warning_count = 0
             if fieldname == 'rn_citation_data':
-                m1_bibs = [element['bibcode'] for element in m1[fieldname]]
-                m2_bibs = [element['bibcode'] for element in m2[fieldname]]
+                m1_bibs = [element['bibcode'] for element in v1]
+                m2_bibs = [element['bibcode'] for element in v2]
                 for v in m1_bibs:
                     if v not in m2_bibs:
                         warning_count += 1
-                        print 'warning: list element missing', v
+                        metrics_logger.warn('{} {} list element missing from second list: {}'.format(bibcode, fieldname, v))
```

Comment:
*iff* the list of compared values is large, then this code will do lots of comparisons (it is n^2; four times)

it should really be modified to do something like
```
v1 = set(t1['key'])
v2 = set(t2['key'])
if v1.differs(v2):
   ....
```

note: i'm assuming here that python 'set' is handling these situations well. It should be, but I didn't test it ;-)